### PR TITLE
Modify parse_type function to use pycparser.CParser parse method with type expressions and add tests to it.

### PR DESCRIPTION
**Pull request by saullocarvalho**
…ruct_deduplication into 'main' function

### CODE REVIEW

File: angr/sim_type.py [hunk 1]
```diff
@@ -1079,19 +1097,43 @@ def parse_type(defn, preprocess=True):
...
-    defn = 'typedef ' + defn.strip('; \n\t\r') + ' QQQQ;'
+    defn = re.sub(r"/\*.*?\*/", r"", defn)
 
-    if preprocess:
-        defn = do_preprocess(defn)
+    parser = pycparser.CParser()
+
+    # pycparser hack to parse type expressions
+    errorlog = logging.getLogger('yacc logger')
+    errorlog.setLevel(logging.ERROR)
```
Comment (rhelmot) [inline angr/sim_type.py]: users generally consider it bad when there are loggers they can't control the level of. I would stick the declaration and setlevel of errorlog in the top scope, and probably change the name to `__name__ + ".yacc"`

File: angr/sim_type.py [hunk 2]
```diff
@@ -1079,19 +1097,43 @@ def parse_type(defn, preprocess=True):
...
-    decl = node.ext[-1].type
+    decl = node.type
     return _decl_to_type(decl)
 
+
+def _accepts_scope_stack():
+    """
+    pycparser hack to include scope_stack as parameter in CParser parse method
+    """
+    def parse(self, text, scope_stack=[dict()], filename='', debuglevel=0):
```
Comment (rhelmot) [inline angr/sim_type.py]: The values that are created when you write expressions as default arguments are evaluated at declaration time, not call time, so this mutable value will be stored by reference as the default argument for all calls, so future calls will receive a mutated default value. The correct way to do things like this is `scope_stack=None`, and then `self._scope_stack = [dict()] if scope_stack is None else scope_stack`.

File: tests/test_types.py [hunk 3]
```diff
@@ -59,7 +59,47 @@ def test_struct_deduplication():
     dhdr = angr.types.define_struct('struct dhdr { struct ahdr a; struct bhdr b; struct chdr c;}')
     assert dhdr.fields['a'].fields
 
+def test_parse_type():
+    int_ptr = angr.types.parse_type('int *')
```
Comment (rhelmot) [inline tests/test_types.py]: can there be a test that runs parse_type with a name involved, like `int *x`?

Comment (rhelmot) [reply tests/test_types.py]: the other major thing missing representation in this test is function pointers

File: angr/sim_type.py (hunk 1 above)
Comment (saullocarvalho) [reply angr/sim_type.py]: I have declared the errorlog and have set its level in the top scope.

File: angr/sim_type.py (hunk 2 above)
Comment (saullocarvalho) [reply angr/sim_type.py]: I have corrected this. Thanks for the information!

File: tests/test_types.py (hunk 3 above)
Comment (saullocarvalho) [reply tests/test_types.py]: I have added tests with function pointer type expression and variable declaration.

**Comment by rhelmot** [comment]
Looks great, thanks!!

**Pull request merged**